mcs/class/corlib/Microsoft.Win32/Win32RegistryApi.cs

```diff
@@ -241,48 +241,66 @@ public object GetValue (RegistryKey rkey, string name, object defaultValue, Regi
 		public void SetValue (RegistryKey rkey, string name, object value, RegistryValueKind valueKind)
 		{
 			Type type = value.GetType ();
-			int result;
 			IntPtr handle = GetHandle (rkey);
 
-			if (valueKind == RegistryValueKind.QWord && type == typeof (long)) {
-				long rawValue = (long)value;
-				result = RegSetValueEx (handle, name, IntPtr.Zero, RegistryValueKind.QWord, ref rawValue, Int64ByteSize); 
-			} else if (valueKind == RegistryValueKind.DWord && type == typeof (int)) {
-				int rawValue = (int)value;
-				result = RegSetValueEx (handle, name, IntPtr.Zero, RegistryValueKind.DWord, ref rawValue, Int32ByteSize); 
-			} else if (valueKind == RegistryValueKind.Binary && type == typeof (byte[])) {
-				byte[] rawValue = (byte[]) value;
-				result = RegSetValueEx (handle, name, IntPtr.Zero, RegistryValueKind.Binary, rawValue, rawValue.Length);
-			} else if (valueKind == RegistryValueKind.MultiString && type == typeof (string[])) {
-				string[] vals = (string[]) value;
-				StringBuilder fullStringValue = new StringBuilder ();
-				foreach (string v in vals)
-				{
-					fullStringValue.Append (v);
-					fullStringValue.Append ('\0');
+			switch (valueKind) {
+			case RegistryValueKind.QWord:
+				try {
+					long rawValue = Convert.ToInt64 (value);
+					CheckResult (RegSetValueEx (handle, name, IntPtr.Zero, RegistryValueKind.QWord, ref rawValue, Int64ByteSize));
+					return;
+				} catch (OverflowException) {
```

Comment:
If `value` is a string then this won't catch the `FormatException`. Maybe just do the try/catch around `Convert.ToInt64` and simply catch all exceptions.